src/librustc_typeck/check/compare_method.rs

```diff
@@ -138,14 +138,51 @@ pub fn compare_impl_method<'a, 'tcx>(ccx: &CrateCtxt<'a, 'tcx>,
     }
 
     if impl_m.fty.sig.0.inputs.len() != trait_m.fty.sig.0.inputs.len() {
-        span_err!(tcx.sess, impl_m_span, E0050,
+        let trait_number_args = trait_m.fty.sig.0.inputs.len();
+        let impl_number_args = impl_m.fty.sig.0.inputs.len();
+        let impl_m_node_id = tcx.map.as_local_node_id(impl_m.def_id).unwrap();
+        let mspan = match tcx.map.expect_impl_item(impl_m_node_id).node {
+            ImplItemKind::Method(ref impl_m_sig, _) => {
+                if let Some(arg) = impl_m_sig.decl.inputs.get(
+                    if trait_number_args > 0 {
+                        trait_number_args - 1
+                    } else {
+                        0
+                    }) {
+                    arg.pat.span
+                } else {
+                    impl_m_span
+                }
+            }
+            _ => bug!("{:?} is not a method", impl_m)
+        };
+        let span = impl_m_span;
+        struct_span_err!(tcx.sess, mspan, E0050,
             "method `{}` has {} parameter{} \
              but the declaration in trait `{}` has {}",
             trait_m.name,
-            impl_m.fty.sig.0.inputs.len(),
-            if impl_m.fty.sig.0.inputs.len() == 1 {""} else {"s"},
+            impl_number_args,
+            if impl_number_args == 1 {""} else {"s"},
             tcx.item_path_str(trait_m.def_id),
-            trait_m.fty.sig.0.inputs.len());
+            trait_number_args)
+            .span_label(trait_item_span.unwrap(),
+                        &format!("trait requires {}",
+                                 &if trait_number_args != 1 {
+                                     format!("{} parameters", trait_number_args)
+                                 } else {
+                                     format!("{} parameter", trait_number_args)
+                                 }))
+            .span_label(span,
```

Comment:
Yeah looks like you just need:

```
-            .span_label(span,
+            .span_label(mspan,
```

And remove the unused `span`.  With that it's at least closer:

```
error[E0050]: method `bar` has 1 parameter but the declaration in trait `Foo::bar` has 4
  --> src/test/compile-fail/E0050.rs:22:5
   |
13 |     fn bar(&self, x: u8, y: u8, z: u8); //~ NOTE trait requires 4 parameters
   |     ----------------------------------- trait requires 4 parameters
...
22 |     fn bar(&self) { } //~ ERROR E0050
   |     ^^^^^^^^^^^^^^^^^ expected 4 parameters, found 1

error[E0050]: method `less` has 4 parameters but the declaration in trait `Foo::less` has 1
  --> src/test/compile-fail/E0050.rs:24:13
   |
14 |     fn less(&self); //~ NOTE trait requires 1 parameter
   |     --------------- trait requires 1 parameter
...
24 |     fn less(&self, x: u8, y: u8, z: u8) { } //~ ERROR E0050
   |             ^^^^^ expected 1 parameter, found 4
```

Not sure why it's doing the full underline rather than just the parameter...
```